draft-ietf-quic-tls.md

```diff
@@ -1197,6 +1197,64 @@ TLS ClientHello.  The server MAY retain these packets for later decryption in
 anticipation of receiving a ClientHello.
 
 
+## Retry Packet Integrity {#retry-integrity}
+
+Retry packets (see the Retry Packet section of {{QUIC-TRANSPORT}}) carry a
+Retry Integrity Tag that provides two properties: it allows discarding
+packets that have accidentally been corrupted by the network, and it ensures
+that valid Retry packets cannot be sent by off-path attackers.
+
+The Retry Integrity Tag is a 128-bit field that is computed as the output of
+AEAD_AES_128_GCM used with the following inputs:
+
+- The key is 128 bits all set to zero.
+- The nonce is 96 bits all set to zero.
+- The plaintext is empty.
+- The associated data is the contents of the Retry Pseudo-Packet, as described
+  in {{retry-pseudo}}:
+
+~~~
+ 0                   1                   2                   3
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+
+|1|1| 3 | Unused|
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                         Version (32)                          |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+| DCID Len (8)  |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|               Destination Connection ID (0..160)            ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+| SCID Len (8)  |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                 Source Connection ID (0..160)               ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                        Retry Token (*)                      ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+| ODCID Len (8) |
```

Comment:
Why do we need the length here?

Reply:
We need the length, as if it's absent, an attacker on the path from client to server but not on the opposite direction can obtain an Initial carrying DCID=X, race an Initial carrying DCID=`prefix || X`. Assuming that the Retry in response to the spoofed Initial wins the race, the client would learn that the value of the token is `correct_token  || prefix`.

That said, I do not think having ODCID Len at this position prevents the attack, because the Retry Token field preceding this field does not have an explicit length. Rather, I think that we need to either place ODCID Len and ODCID before the Retry packet, or add the length of the Retry packet at the very front.

Reply:
Well, if the token is followed by an integrity tag, the would implicitly define the token length, no?

Reply:
@mikkelfj This is about the format of the pseudo packet that we use to *calculate* the integrity tag.

Reply:
Good catch, moved ODCID to the start to remove the ambiguity

Reply:
@kazuho I would assume it had a pseudo packet length use to compute the tag, otherwise it would be a futile exercise. I'm probably still missing something. A weaker integrity check would need some padding protection but GCM should be fine.

Reply:
yeah, I'm missing that two fields without a length is ambigious.